src/System.Reflection.Context/src/System/Reflection/Context/Virtual/VirtualProperty.PropertySetter.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Reflection.Context.Custom;
+
+namespace System.Reflection.Context.Virtual
+{
+    partial class VirtualPropertyInfo
+    {
+        private class PropertySetter : PropertySetterBase
+        {
+            private readonly Action<object, object> _setter;
+            private readonly ParameterInfo _valueParameter;
+            private readonly IEnumerable<Attribute> _attributes;
+
+            public PropertySetter(VirtualPropertyBase property, Action<object, object> setter, IEnumerable<Attribute> setterAttributes)
+                : base(property)
+            {
+                Debug.Assert(null != setter);
+
+                _setter = setter;
+                _valueParameter = new VirtualParameter(this, property.PropertyType, "value", 0);
+                _attributes = setterAttributes ?? CollectionServices.Empty<Attribute>();
+            }
+
+            public override ParameterInfo[] GetParameters()
+            {
+                return new ParameterInfo[] { _valueParameter };
+            }
+
+            public override object Invoke(object obj, BindingFlags invokeAttr, Binder binder, object[] parameters, CultureInfo culture)
+            {
+                // invokeAttr, binder, and culture are ignored, similar to what runtime reflection does with the default binder.
+
+                if (parameters == null || parameters.Length != 1)
+                    throw new TargetParameterCountException();
+
+                object value = parameters[0];
+
+                if (obj == null)
+                    throw new TargetException(SR.Target_InstanceMethodRequiresTarget);
+
+                if (!ReflectedType.IsInstanceOfType(obj))
+                    throw new TargetException(SR.Target_ObjectTargetMismatch);
+
+                if (ReturnType.IsInstanceOfType(value))
```

Comment:
This checks seems unnecessary as ReturnType here is always typeof(void) and without bad IL hacks you can't initialize System.Void. As I couldn't find a test path for it I tend to remove it.

cc @stephentoub @jkotas

Reply:
How do you know that ReturnType is always typeof(void) here?

I'd suggest leaving it for now.

Reply:
If you step into how ReturnType is defined for PropertySetter you see that it always returns typeof(void). But doesn't really matter as this package probably won't be used in perf critical scenarios.